### ISSUE TITLE
Identical, anonymous intersection types causes incompatible import [assignment] error when imported via *

### ISSUE THREAD

**Issue by noahbkim** [python/mypy#16262]
**Bug Report**

If you import `*` from two modules that infer an identical intersection type, mypy will report an error along the lines of 

```
Incompatible import of "<intersection>" (imported name has type "type[a.<intersection>]", local name has type "type[b.<intersection>]")  [assignment]
```

**To Reproduce**

This repro requires multiple files. They are delineated by comments.

```python
# shared.py
class Foo: ...
class Bar: ...

# a.py
from shared import Foo, Bar
a_foos: list[Foo]
a_bars = [f for f in a_foos if isinstance(f, Bar)]

# b.py
from shared import Foo, Bar
b_foos: list[Foo]
b_bars = [f for f in b_foos if isinstance(f, Bar)]

# c.py
from a import *
from b import *
```

The offending `mypy` invocation is as follows:

```
$ mypy -m c
```

**Expected Behavior**

This should not emit an error because no runtime locals are actually assigned, or at the very least, because the intersections are identical.

**Actual Behavior**

```
c.py:2: error: Incompatible import of "<subclass of "Foo" and "Bar">" (imported name has type "type[b.<subclass of "Foo" and "Bar">]", local name has type "type[a.<subclass of "Foo" and "Bar">]")  [assignment]
Found 1 error in 1 file (checked 1 source file)
```

**Your Environment**

This bug was reproduced in a fresh venv on Debian GNU/Linux 10 (buster):

```
$ python --version
Python 3.10.13
$ mypy --version
mypy 1.6.0 (compiled: yes)
```

**Comment by oremanj** [comment]
Note that the same error occurs even if `a.py` and `b.py` use `__all__` to limit what `from .. import *` imports from them. The apparently internal/synthesized type still gets imported by the `import *` even though it's not named in `__all__`.
